### ISSUE TITLE
[Web] Add Functionality Features to Model Versioning

### ISSUE THREAD

**Issue by StephenODea54** [wetherc/dstk#63]
Server actions to include in the model versioning portion of the app

- [ ] Sorting
- [ ] Filtering
- [ ] Pagination
- [ ] Create New Model UI/UX and Functionality
  - [ ] Create Model Version
  - [ ] Upload Files (Multipart upload for each file)
  - [ ] Publish Model Version

**Comment by StephenODea54** [comment]
@wetherc Read through #61 again. Just to double check, all we need the user to provide is an optional description and the actual files to be uploaded, right?

**Comment by wetherc** [comment]
Yup exactly. Right now it's restricted to a single file. A couple iterations down the road (so like, next week) I'll fix things up a bit so we have good support for an arbitrary number of files

**Comment by wetherc** [comment]
For the next iteration, the model version workflow will change to "Create model version" > upload file(s) > "finalize model version" all as discrete user actions if that impacts what you're thinking of for the frontend.

**Comment by StephenODea54** [comment]
Gotcha, thanks! I was just trying to mock up what the frontend looks like for this process. The difference between being able to upload a single file or multiple won't change what that specific UI piece looks like, just the API calls it will be making based on the future iterations.

What are we thinking for the flow for the user actions? I was thinking:
- Click create a new version, have user go to new page
- Have a single form the user fills out (description + file upload(s))
- Upon successful submission, change the view to display a progress bar that shows the succession between steps (model version was created, connecting w/ your storage provider, uploading files in X steps, etc.)
- Display a notification that everything went hunky-dory and return the user to the model version page
- If anything goes wrong, give the user a notification of the error and revert any steps currently completed.

Based on what you said about the future iterations, it sounds like the workflow might be better suited with a [multi-step form](https://blog.hubspot.com/marketing/multi-step-forms)

**Comment by StephenODea54** [comment]
Kind of what I was getting at w/ my last comment:

Here's a UI mockup if everything were done on a single page within a single form:

<img width="1039" alt="image" src="https://github.com/wetherc/dstk/assets/77359138/0728a3fe-7e3f-4ed5-96a8-6aba5bcb58a6">

The Model Name and Model Version are disabled inputs. They are there just to show what the finalized values will be. Was thinking about putting Storage Provider information in as well, but don't know if that's beneficial for the user to know.

**Comment by wetherc** [comment]
I like it!

Just thinking about the API methods we're exposing, it might make sense to split it across 3 user actions still.

On that page as mocked up, we won't know what the new version's numeric identifier is until it's been created. (You could `listVersionsForModel()` it on the frontend and do it yourself, but then you're kind of guessing at database state and hoping it doesn't change between the time you render that screen and the time the user hits Submit.

If we have it split into discrete steps for:

  - Create Model Version
  - Upload Files
  - Publish Model Version

then that will align with the API methods pretty cleanly (I'm going to totally separate the S3 operations from the model version mutations; right now they're kind of commingled when they don't need to be). It also gives the user flexibility to continue to add files to their model version if they don't have everything exactly ready right then and there.

What're your thoughts?

**Comment by StephenODea54** [comment]
Yeah no that totally makes sense let's do that.

For the three discrete steps we'll split the workflow into:

- **Create Model Version**: Users provides the optional description. Makes API call to the `CreateModelVersionMutation()` (We'll already have the Model ID from the page URL).
- **Upload Files**
  - User can upload file(s).
  - Will make a call to whatever the S3 operation mutation looks like.
- **Publish Model Version**
  - Will show the user the new model version and whatever else we want to show them before finalizing.
  - I assume this step will flip the `isFinalized` indicator once the user hits submit.

**Comment by wetherc** [comment]
Just for reference, the current version of #61 now cleanly supports multi-file upload. The workflow is to create a multipart upload for each file, and then for each part of that file, request a pre-signed URL to post the file contents to

**Comment by StephenODea54** [comment]
Step 1

<img width="1025" alt="image" src="https://github.com/wetherc/dstk/assets/77359138/88a90ec4-7031-423e-b184-c46e3cff9d01">

**Comment by wetherc** [comment]
Ooh I like

Question: what happens if the user exits the flow partway through? Can they resume later?

**Comment by StephenODea54** [comment]
That's the reason why I haven't submitted a PR yet. I'm still thinking about how to save the form state between steps on the client. Normally in a multi-step form the different inputs are broken out into discrete steps but there is only one submission. Our case is different because our workflow has API calls that need to be validated in between each step. 

Using the first part as an example:
1. User maybe submits description.
2. GraphQL mutation is ran
3. Use Apollo Client to validate submission state
    - Is it still submitting? Display some type of loading indicator
    - Did it fail? Display a notification to the user
    - Did it succeed? Great, update some type of local state that says "hey this step is good to go, generate new form and ui updates for the file upload part"

The tricky part is how do we save this data such that the user:
  - Wants to go to a previous step and make edits?
  - Goes on lunch break and wants to save their progress for later?

Right now, we are using no auth so if you leave the tab open everything is saved in local state, but
1. That's a janky way to save progress... even though I'm totally guilty of doing this lol
2. When we implement auth, their session token can expire and if the steps aren't saved anywhere they'll lose their progress when they log back in. This can cause issues where, let's say, they complete step 1 but then lose their progress. That means we'll have a random model version floating around with no other information attached to it.

TLDR; No ability for user to save now but 100% will implement once I kind of have a better idea on how to save the state.

**Comment by wetherc** [comment]
Do you need to save form state? If a version exists that isn't published, the user can resume adding files at any point. If a version exists that is published, it's in read-only mode. If a user closes out the page, no partial state gets saved

The only thing we *might* want to add is the ability to edit the description of a model version

**Comment by wetherc** [comment]
(The point being, the only state I'd argue that we care about is what already gets persisted to the db through the normal course of invoking the existing mutations)

**Comment by StephenODea54** [comment]
How I have it configured now, yeah I need to save form state because it's all bundled into one parent component w/ one entry point (the create new model version on the model versioning page). So it needs some type of state saved to know which steps are completed, what step the user is on, etc.

Now what we could do is separate the actions into three separate components w/ three possible entry points:
- Create Model Version
  - Can access through the model versioning page by clicking the Create Model Version Button
- Upload Files
  - Can access through somewhere else, like having a link in the table on the model versioning page that is only accessible to models that are not published.
  - You could also add in things like: upon successful completion of step 1, could provide the user w/ a link or notification that says "your model version was successfully registered, want to upload stuff right now? Well, click here!"
- Publish
  - Kind of similar logic to the upload files step.
  - Have a button somewhere (or a row click) on a certain version that takes you to the "published" model version page.
    - If it's published, it is read only
    - If it's not, well then you can review and submit stuff!

So actually, after typing this out, I think I'm totally going to go this route because it makes the frontend implementations simpler lol.

Also, if my rambling made no sense, I am definitely okay with syncing on this.

**Comment by wetherc** [comment]
Ohhhh gotcha. In that case yeah, piecing it out into 3 separate components sounds like the way to go. Each of those 3 actions is decoupled from the other two, so reflecting that in how the components are structured probably keeps life simple.

Thinking of it from a routing POV, it would be a `/models/<model>/versions/create` endpoint, a `versions/<version>/upload` endpoint, and a `versions/<version>/publish` endpoint where each exists roughly independently of the other two

**Comment by StephenODea54** [comment]
Yes! I love it. I'll get right on it 🫡

**Comment by wetherc** [comment]
> Yes! I love it. I'll get right on it 🫡

Be careful, Cynthia is always watching 👀 (kidding, you know we love you Cynthia!)

**Comment by StephenODea54** [comment]
> > Yes! I love it. I'll get right on it 🫡
> 
> 
> 
> Be careful, Cynthia is always watching 👀 (kidding, you know we love you Cynthia!)

All love Cynthia 😂 ❤️

**Comment by StephenODea54** [comment]
Moving to `dstk-web` repo
